scRNA-seq-advanced/02-celltype_assignment.Rmd

```diff
@@ -0,0 +1,492 @@
+```
+
+
+Read in the SingleCellExperiment
+```{r}
+sce <- readRDS(sce_file)
+```
+
+
+Let's explore what is in the SCE already:
+
+```{r}
+sce
+```
+
+Notice the `altExpNames`?
+Those are the ADT data, stored as an _Alternative Experiment_.
+To access these, we can use the `altExp()` function. 
+Let's look at what we have in that slot:
+
+```{r}
+altExp(sce, "ADT")
+```
+
+
+Ah! Another SingleCellExperiment! Inception!
+Since there is only one `altExp`, the second argument is optional, so we won't use it for the rest of this notebook.
+Here the columns are again the cells, but now the rows are the proteins that were assayed with the antibodies and tags.
+
+```{r}
+# What proteins were assayed?
+adt_tags <- rownames(altExp(sce))
+adt_tags
+```
+
+
+### Clustering redux
+
+These data don't have clusters assigned yet, so let's add those, using the PCA matrix we have for the gene expression data.
+
+
+```{r}
+# perform clustering
+nn_k <- 20
+nn_clusters <- bluster::clusterRows(
+  reducedDim(sce, "PCA"), 
+  bluster::NNGraphParam(
+    k = nn_k, 
+    type = "jaccard",
+    cluster.fun = "louvain"
+  )
+)
+
+# add clusters to colData
+sce$nn_cluster <- nn_clusters
+```
+
+```{r}
+scater::plotUMAP(sce, 
+                 colour_by = "nn_cluster")
+```
+But what are these clusters? Do they correspond to cell types?
+
+## Cell surface markers
+
+The markers we have were presumably chosen for the fact that they mark particular cell types, so let's look at a few.
+
+CD3 is a marker of T-cells 
+
+```{r}
+scater::plotUMAP(sce, 
+                 colour_by = "CD3")
+```
+
+CD4+ T-cells
+
+```{r}
+scater::plotUMAP(sce, 
+                 colour_by = "CD4")
+```
+CD8+ T-cells
+
+```{r}
+scater::plotUMAP(sce, 
+                 colour_by = "CD8")
+```
+
+
+So one thing we could do is construct a set of rules based on ADT tags, similar to how you might set gating rules for cell sorting.
+
+
+What thresholds should we use?
+If this is a useful marker, we should see a bimodal plot of the marker density.
+
+We'll do a bit of wrangling to pull out the ADT data, transpose it so each row is a cell,
+then convert it into a data frame for plotting. 
+(We have to first convert to a "regular" R matrix or R won't know how to convert it to a data frame)
+
+```{r}
+# convert to a data frame
+adt_df <- logcounts(altExp(sce)) %>%
+  t() %>% 
+  as.matrix() %>%
+  as.data.frame()
+```
+
+With a bit more wrangling, we can put the expression data for the three tags we care about into a single column, which will allow us to plot all three together a bit more neatly.
+Here we are using the `tidyr::pivot_longer()` function, which allows us to convert our data frame with one column per tag into a data frame with separate columns for the tag id (`ADT`) and the expression value (`logcount`).
+
+```{r}
+adt_df_long <- adt_df %>%
+  dplyr::select(c("CD3", "CD4", "CD8")) %>%
+  tidyr::pivot_longer(everything(), names_to = "ADT", values_to = "logcount")
```

Comment:
```suggestion
  tidyr::pivot_longer(dplyr::everything(), names_to = "ADT", values_to = "logcount")
```

Reply:
I don't like this suggestion? `everything()` comes from `tidyselect` and I really don't want to get into that... For these helper functions I think we do fine with not being fully explicit... It is similar to how I don't use `dplyr::arrange(dplyr::desc(x))`

Reply:
The function is exported to `dplyr` from `tidyselect`, and I tend not use the latter for this kind of namespacing _when teaching_ since it's not a core tidyverse package and it can be strange for newcomers. Yes it is true that `tidyselect` is "more" right. 

But I think my main point is...how will this code be able to run otherwise? `everything()` isn't in the environment.

Reply:
It is in the calling environment within the function... so a bit of magic, but it works.

Reply:
That _is_ magic! I'm realizing I rarely use `::` w/ core tidyverse on my own unless I'm writing a package where `::` is needed to appease CRAN checks. Punchline, I'm extremely comfortable with you ignoring my suggestion in this case!